### ISSUE TITLE
fatal error: concurrent map writes with loki-stack-2.5.0

### ISSUE THREAD

**Issue by Nexus2k** [grafana/loki#4789]
**Describe the bug**
Since we've enabled autoscaling on our GCP k8s cluster we're seeing an increased amount of crashing promtail pods.
The error is `fatal error: concurrent map writes`. iiuc this did happen before in https://github.com/grafana/loki/issues/2456 and https://github.com/grafana/loki/issues/2569 so might be a regression?

**To Reproduce**
1. Enabled Autoscaling on GCP
2. Deployed loki-stack-2.5.0 helm chart (w/o any resource limits)
3. crashloops happen

**Expected behavior**
pods not crashing

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: helm using loki-stack-2.5.0 chart (afaik loki v2.1.0)

**Screenshots, Promtail config, or terminal output**
Output of `kubectl logs loki-promtail-fgf84`:
```
fatal error: concurrent map writes

goroutine 454 [running]:
runtime.throw(0x2a813db, 0x15)
	/usr/local/go/src/runtime/panic.go:1116 +0x72 fp=0xc000d23800 sp=0xc000d237d0 pc=0x43adf2
runtime.mapassign_fast64(0x25e6e20, 0xc0009f58c0, 0x246b2d5711eb53b1, 0xc001644780)
	/usr/local/go/src/runtime/map_fast64.go:101 +0x33e fp=0xc000d23840 sp=0xc000d23800 pc=0x41641e
github.com/grafana/loki/pkg/logql/log.(*BaseLabelsBuilder).ForLabels(0xc000a14380, 0xc000281960, 0x7, 0x7, 0x246b2d5711eb53b1, 0x400)
	/src/loki/pkg/logql/log/labels.go:116 +0x1e5 fp=0xc000d238a0 sp=0xc000d23840 pc=0xcd31e5
github.com/grafana/loki/pkg/logql/log.(*pipeline).ForStream(0xc0009f5920, 0xc000281960, 0x7, 0x7, 0xc001ac7708, 0xc000dd2690)
	/src/loki/pkg/logql/log/pipeline.go:118 +0x153 fp=0xc000d23938 sp=0xc000d238a0 pc=0xcd89b3
github.com/grafana/loki/pkg/logentry/stages.(*matcherStage).Process(0xc0009d5bc0, 0xc0009f5d10, 0xc0009f5d40, 0xc000b60720, 0xc000865a50)
	/src/loki/pkg/logentry/stages/match.go:135 +0x138 fp=0xc000d23b10 sp=0xc000d23938 pc=0xd21e38
github.com/grafana/loki/pkg/logentry/stages.(*Pipeline).Process(0xc000a33680, 0xc0009f5d10, 0xc0009f5d40, 0xc000b60720, 0xc000865a50)
	/src/loki/pkg/logentry/stages/pipeline.go:109 +0x213 fp=0xc000d23c78 sp=0xc000d23b10 pc=0xd26b53
github.com/grafana/loki/pkg/logentry/stages.(*Pipeline).Wrap.func1(0xc0009f5d10, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0x0, 0xffffffffffffffff)
	/src/loki/pkg/logentry/stages/pipeline.go:129 +0x110 fp=0xc000d23d08 sp=0xc000d23c78 pc=0xd2ec50
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc000a37fc0, 0xc0009f5d10, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0x203000, 0xc0009f5ce0)
	/src/loki/pkg/promtail/api/types.go:24 +0x65 fp=0xc000d23d58 sp=0xc000d23d08 pc=0xd017a5
github.com/grafana/loki/pkg/promtail/api.AddLabelsMiddleware.func1.1(0xc0009f5ce0, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0x0, 0x0)
	/src/loki/pkg/promtail/api/types.go:45 +0xa3 fp=0xc000d23dc0 sp=0xc000d23d58 pc=0xd01983
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc000f2f2a0, 0xc0009f5ce0, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0xc000d23e58, 0x4)
	/src/loki/pkg/promtail/api/types.go:24 +0x65 fp=0xc000d23e10 sp=0xc000d23dc0 pc=0xd017a5
github.com/grafana/loki/pkg/promtail/api.AddLabelsMiddleware.func1.1(0xc0009f5cb0, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0x2941bc0, 0xc0009f5cb0)
	/src/loki/pkg/promtail/api/types.go:45 +0xa3 fp=0xc000d23e78 sp=0xc000d23e10 pc=0xd01983
github.com/grafana/loki/pkg/promtail/api.EntryHandlerFunc.Handle(0xc000f2f300, 0xc0009f5cb0, 0xc05df5cc53e97cb9, 0x14258f91c, 0x413f820, 0xc000978120, 0x86, 0x0, 0x0)
	/src/loki/pkg/promtail/api/types.go:24 +0x65 fp=0xc000d23ec8 sp=0xc000d23e78 pc=0xd017a5
github.com/grafana/loki/pkg/promtail/targets/file.(*tailer).readLines(0xc000fa4680)
	/src/loki/pkg/promtail/targets/file/tailer.go:150 +0x57c fp=0xc000d23fd8 sp=0xc000d23ec8 pc=0x205599c
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc000d23fe0 sp=0xc000d23fd8 pc=0x473a01
created by github.com/grafana/loki/pkg/promtail/targets/file.newTailer
	/src/loki/pkg/promtail/targets/file/tailer.go:80 +0x53d
```

Full stack trace available here: https://gist.github.com/Nexus2k/da68c76b0b1eb663b073c6b723a0644e

**Comment by Nexus2k** [comment]
Please let me know if I can provide any additional information to get this fixed.

**Comment by cyriltovena** [comment]
Hello,

This is known bug that has been fixed in 2.2 and forward.

Thanks

Le lun. 29 nov. 2021 à 14:23, Philipp D. ***@***.***> a
écrit :
